### DISCUSSION TITLE
Http endpoint to change scrape_interval and scrape_timeout.

### DISCUSSION THREAD

**Discussion by VenkateswaranJ** [prometheus/prometheus#10515] [Q&A] [answered]
Currently, there is no Http endpoint for changing scrape_interval and scrape_timeout. One has to edit the config file manually and reload the config every time. Also in [<http_sd_config>](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#http_sd_config) it is only possible to add target URLs and labels but it's not possible to add scrape_interval for each target.

**Comment by roidelapluie** [comment, marked as answer]
With http_sd_config you can set scrape interval for each target:
```
[
    {
        "targets": ["10.0.40.3:9100"],
        "labels": {
            "__meta_datacenter": "london",
            "__scrape_interval__": "1m",
            "__scrape_timeout__": "5m"
        }
    },
    {
        "targets": ["10.0.40.1:9100"],
        "labels": {
            "__meta_datacenter": "london",
            "__scrape_interval__": "1m",
            "__scrape_timeout__": "5m"
        }
    }
]

```

You can also do this with relabel_configs. Without knowing more about your usecase, it is difficult to help you more.

**Reply by VenkateswaranJ** [reply]
@roidelapluie Thanks a lot for your answer.
Is there any way to change scrape_intervel of global config via http call?

> Without knowing more about your use case, it is difficult to help you more.

I don't have any special use case. I have a server that runs Prometheus and it gets connected to multiple client machines which run different types of exporters. When a new client connects to the server I just add the exporter url of the client to Prometheus via the <http_sd_config> endpoint. I have a frontend UI that connects to the server and sees the metrics. Now the UI has a requirement that it needs to change the scrape_interval (globally) as well as on a per-target basis. Now I got the answer for how to change per-target basis and would like to know it's possible to change the global config via http (instead of changing the config file)

**Reply by VenkateswaranJ** [reply]
@roidelapluie Unfortunately the answer you suggested is not working. Below is my setup.

Prometheus version: v2.34.0

Config file: (trying with a single target)
```
global:
    scrape_interval: 15s

# Here it's Prometheus itself.
scrape_configs:
    - job_name: 'prometheus'
      static_configs:
          - targets: ['localhost:9090']
    - job_name: 'httpsd'
      http_sd_configs:
          - url: "http://host.testcontainers.internal:8094/targets"
            refresh_interval: 5s
```

Target list JSON:

```
[
  {
    "targets": [
      "node:9100"
    ],
    "labels": {
      "hostId": "b3ed4d66-2da4-47b7-a330-1d1e81b00727",
      "hostType": "Rsp1232HL",
      "__scrape_interval__": "5s"
    }
  }
]
```

As you can see above that I have set __scrape_interval__ to 5s and I also confirmed that Prometheus calls the http_sd endpoint every refresh interval (5s) but I don't see the target in Prometheus web-ui. **If I remove the `__scrape_interval__` from the target list JSON, everything works fine and the target takes scrape_interval from the global config(15s).**

I see in the Prometheus document that it's an experimental feature. Is it anything to do with that?

> The __scrape_interval__ and __scrape_timeout__ labels are set to the target's interval and timeout. This is experimental and could change in the future.

Is there any other way I can configure scrape_inerval of discovered targets?

**Comment by VenkateswaranJ** [comment]
Further findings, I increased __scrape_interval__ of targets from 5s to 10s and it works. So there might be a relation between global scrape interval and per target scrape interval. Because in the above case my global scrape_interval is 15s and if I set the target scrape_interval less than 10s then the Prometheus does not discover the target but everything works if I increase the scrape_interval to 10s or above.

**Reply by roidelapluie** [reply]
I suspect you also need to lower your scrape_timeout. Did you check the logs?

**Reply by VenkateswaranJ** [reply]
@roidelapluie yes, you are right.
Scrape_interval should be greater than or equal to Scrape_timeout.
